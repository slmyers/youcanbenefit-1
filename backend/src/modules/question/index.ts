export { QuestionModule } from "./question.module"