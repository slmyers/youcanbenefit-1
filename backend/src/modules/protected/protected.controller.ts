import { Get, Controller, Param, Body, Post, Put, Delete, Req } from '@nestjs/common';
import { Observable } from "rxjs/Observable"
import { ProgramService } from "../Program/program.service";
import { ApplicationProgramDto } from "./AplicationProgram.dto";
import { ApplicationQueryService } from "../query/ApplicationQuery.service";
import { ApplicationQueryDto } from "../query/ApplicationQuery.dto"
import "rxjs/add/observable/fromPromise"
import "rxjs/add/observable/from"
import "rxjs/add/operator/map"
import "rxjs/add/observable/zip"
import "rxjs/add/operator/mergeMap"
import "rxjs/add/observable/throw"
import "rxjs/add/observable/of"
import "rxjs/add/operator/catch"
import { QuestionService } from '../question/question.service';
import { QuestionDto } from '../question/question.dto';
import {ScreenerDto} from "../screener/screener.dto";
import {ScreenerService} from "../screener/screener.service";
import { ConstantsReadonly } from "../constants.readonly"
const fs = require('fs');
const path = require('path');


@Controller('protected')
export class ProtectedController {
    constructor(
        private programService: ProgramService,
        private queryService: ApplicationQueryService,
        private questionService: QuestionService,
        private screenerService: ScreenerService
    ) {}

    @Get('/login/')
    loginAlwaysTrue(): Observable<{[key: string]: boolean}> {
        return Observable.of({created: true})
    }

    @Get('/question/')
    getQuestions(): Observable<QuestionDto[]> {
        return this.questionService.getQuestions();
    }

    @Get('/screener/')
    getScreenerWithQuestions(): Observable<any> {
        return Observable.zip(
            this.screenerService.getLatest(),
        ).map( ([screener]) => {
            return {
                ...screener
            }
        })
    }

    @Post('/screener/')
    saveScreener(@Body() data) {
        return Observable.zip(
            this.screenerService.update((<ScreenerDto> data)),
            this.questionService.updateQuestions(data['questions'])
        ).map(([screener, questions]) => {
            return {screener, questions}
        })
    }

    @Get('/program/')
    getProgramsWithQueries(): Observable<ApplicationProgramDto[]> {
        return Observable.zip(
            this.programService.findAll(),
            this.queryService.findAll()
        ).map(([programs, queries]) => {
            return programs.map( program => {
                return new ApplicationProgramDto(
                    queries.filter(query => query.guid === program.guid),
                    program,
                    program.guid,
                );
            })
        })
    }

    @Get('/program/:guid')
    getProgramWithQueries(@Param() params): Observable<ApplicationProgramDto> {
        const guid = params.guid;
        return Observable.zip(
            this.programService.getByGuid(guid),
            this.queryService.getByGuid(guid)
        ).map( ([program, queries]) => {
            return new ApplicationProgramDto(
                queries,
                program,
                program.guid
            )
        })
            .catch(err => {
                console.error(err);
                return Observable.throw(err);
            })

    }

    @Post('/program/')
    createProgramWithQueries(
        @Body("user") user,
        @Body("application") application,
    ): any {
        return Observable.zip(
            this.programService.create(user),
            Observable.from(application)
                .mergeMap( (query: ApplicationQueryDto) => this.queryService.create(query))
                .catch(() => Observable.throw(false))
        )
            .map( ([{created}, queriesCreated]) => {
                return created === true && queriesCreated === true ? { created: true} : { created: false }
            })
    }

    @Put('/program/')
    updateProgramWithQueries(@Body() data: any): any {
        return Observable.zip(
            this.programService.index(data.user),
            Observable.from(data.application).mergeMap((query: ApplicationQueryDto) => this.queryService.index(query))
        )
            .map( ([userUpdated, queriesUpdated]) => {
                return userUpdated.created === true && queriesUpdated.created === true ? { updated: true} : { updated: false }
            })
    }

    @Delete('/program/:guid')
    deleteProgramAndQueries(@Param() params): Observable<any> {
        const guid = params.guid;

        return Observable.zip(
            this.programService.deleteByGuid(guid),
            this.queryService.deleteByGuid(guid)
        )
    }

    @Delete('/query/:id')
    deleteQueryById(@Param() params): Observable<any> {
        return this.queryService.deleteById(params.id)
            .map(res => res.deleted ? {found: true, deleted: true } : { found: null, deleted: false} )
    }

    @Post('/query/')
    updateOrCreateQuery(@Body("query") query, @Body("guid") guid): Promise<any> {
        return this.queryService.index({
            ...query,
            guid
        })
    }

    @Put('/program-description/')
    updateUserFacingProgram(@Body() data): Promise<any> {
        return this.programService.index(data)
    }
}
