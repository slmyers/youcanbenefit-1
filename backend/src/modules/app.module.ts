import { Module, MiddlewareConsumer, RequestMethod } from '@nestjs/common';
import { CorsMiddleware } from '../middleware/cors.middleware'
import { AppController } from './app.controller';
import { ProgramModule } from "./Program"
import { QueryModule } from "./query"
import { ProtectedModule } from "./protected"
import { ApiModule } from "./api"
import { DataModule } from './data/data.module';
import { LogModule } from './log/log.module'
import { ConfigModule } from './config.module'
import { QuestionModule } from './question/question.module';

@Module({
  imports: [
    ConfigModule,
    ProgramModule,
    QueryModule,
    QuestionModule,
    ProtectedModule,
    ApiModule,
    DataModule,
    LogModule,
    QuestionModule
  ],
  controllers: [AppController]
})
export class ApplicationModule {
  configure(consumer: MiddlewareConsumer) {
    consumer.apply(CorsMiddleware).forRoutes({
      path: '*',
      method: RequestMethod.ALL
    })
  }
}
