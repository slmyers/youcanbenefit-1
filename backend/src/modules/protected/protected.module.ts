import { Module } from '@nestjs/common';
import { ProgramModule } from '../Program'
import { QueryModule } from "../query";
import { ProtectedController } from "./protected.controller";
import { QuestionModule } from "../question";
import { ScreenerModule } from "../screener";

@Module({
    imports: [
        ProgramModule,
        QueryModule,
        QuestionModule,
        ScreenerModule
    ],
    controllers: [ ProtectedController ],
    providers: [ ProtectedController ]
})
export class ProtectedModule {}
