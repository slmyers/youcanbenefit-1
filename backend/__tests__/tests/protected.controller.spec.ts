import * as request from 'supertest';
import { Test } from '@nestjs/testing';
import { ProtectedModule } from '../../src/modules/protected'
import { INestApplication } from '@nestjs/common';

describe('/protected', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const module = await Test.createTestingModule({
      imports: [ProtectedModule]
    })
    .compile();
    app = module.createNestApplication();
    await app.init();
  })

  describe('/login', () => {
    it('GET /login will allow you to login, auth is handled by reverse proxy', () => {
      return request(app.getHttpServer())
        .get('/protected/login/')
        .expect(200)
        .expect(res => {
          expect(res.body.created).toBe(true)
        });
    });
  });

  describe('/program -- CRUD with program datastructures', () => {
    it('GET /program will return all programs and their queries', () => {
      return request(app.getHttpServer())
        .get('/protected/program/')
        .expect(200)
        .expect(response => {
            const extractedProperties = response.body.map( applicationProgramDto => applicationProgramDto.guid);
            extractedProperties.sort( (a, b) => a.localeCompare(b));
            
            expect(extractedProperties).toMatchSnapshot();
        })
    });

    it('GET /program/:guid will return a single program and its queries', () => {
      return request(app.getHttpServer())
        .get('/protected/program/hp5mupqcVY8ZMiKg7Q91Uoi4Wf')
        .expect(200)
        .expect(response => {
          const applicationProgramDto = response.body;
          const extractedProperties = {
              guid: applicationProgramDto.guid,
              user: applicationProgramDto.user,
              queryIDs: applicationProgramDto.application.map(query => query.id)
          };
          delete extractedProperties.user.created;
          extractedProperties.queryIDs.sort( (a, b) => a.localeCompare(b));
          expect(extractedProperties).toMatchSnapshot();
        })
    });
  });

  afterAll(async () => {
    await app.close();
  });
});