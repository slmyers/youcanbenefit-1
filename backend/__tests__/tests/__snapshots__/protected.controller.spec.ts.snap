// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/protected /program -- CRUD with program datastructures GET /program will return all programs and their queries 1`] = `
Array [
  "1iXreywnofiF2jKKknlbKRXAtX",
  "1lKYPN8nNaAMldGufGZNEwfjIK",
  "5b1J8A2fhE9cwVo9W5MfbLmuNr",
  "7OFwDmSwGTZO5J3PeLQ85sxLZn",
  "7Vo9xE20WJJJ1ZORDlog1odyzo",
  "8Lqu2z4WCAP9w5sjhcAeOKuHN5",
  "8U2rl4Ha5GXBVqjNStudvzHIch",
  "9dHBCUXa0z8tEZg6QcYMBJVXjf",
  "9rRXexrA3JzmXbFdKCmjtTj57l",
  "BcKJhMTErAsHfXq1jzealos8kp",
  "bgUdObkiCIh0sgQB5019ZJZsl7",
  "bzf4ryrVVsKkqpsEA5dBZd2kP5",
  "eJYKjE0tT13d1kBH9WLbgETu58",
  "ELKEEgllSRMnkqoHN8N2Ufmxaw",
  "EY1QV5T6AyB4Ykl2Lt1Nt9kf9g",
  "GiufEa3ziFwfBEgkjI2gIlfNkk",
  "HiGYI2bhBoQ6n82R9ptoNZ0oaQ",
  "hp5mupqcVY8ZMiKg7Q91Uoi4Wf",
  "iHPbkwKFsck9ZLnMePKSj5e5mz",
  "IO2YsSRkLCIZwIcYGjO6rNkxUE",
  "j1g5dV0buzXSZoNNHborFuLIko",
  "JIcpL1cap5Eg0ikPk2tecp6JG7",
  "Jz15uGLGW9eeopL31U3lJ46LvB",
  "K08iJHYVRGL5pBOomOvwa2zwX2",
  "kc0ov01FtBgr88C5P6EE6lnC8s",
  "KUpG0etZQ91lIlOmSGEVromWec",
  "NMh2Fu1IG5XnpOf5b6zu2t1Hh6",
  "nQXxT1OMhPGbkqkPHc2QTpofRv",
  "NTVwRTvMb82TIB5VLbf3v36tZN",
  "OGR28icdrofstoYETvIGg19XAv",
  "oZ68OrZFpvkMMGHqV3CFELhuVY",
  "sgKMX0336kgepHbPeObRLOvcFL",
  "tNGy4XadOixXFFaxRRTKe8dqNs",
  "tWucjcIELoF0O1Z0DksuXWDGY7",
  "UbdhHxTUe3N5zm5BxR6WQ2i6vf",
  "V3eKEkg2SL4UTFOGd0dvcd2zEO",
  "Y5KEzulkwxviOuzB4Ny5QTzMGD",
  "ZPNzWEr5XfUq5aYdy2bSYIKO5T",
  "zXRcJxGzIrwhmk5LTbE4jdwBMh",
]
`;

exports[`/protected /program -- CRUD with program datastructures GET /program/:guid will return a single program and its queries 1`] = `
Object {
  "guid": "hp5mupqcVY8ZMiKg7Q91Uoi4Wf",
  "queryIDs": Array [
    "C8NZoL1Nf7VEgEC4VMFI9GlebN",
    "d5NswIlUvzxY0AJ1hGm4eKQtS4",
    "fMwdx1jiAZUt09ALDer98GardT",
    "JeXk8cOZnqJMepmHpxDqpsEXNE",
    "jI1HlTlpbULsSVUmHnYi3ZlVJ7",
    "jzMQkEK41gnhmwWyxAuWF3eZ4d",
    "NOyR66aE30dJYXgdHJ2CvGLpnx",
    "qZyokzKnpIw3Ue6mdTi8je3knK",
    "vCYJSMCl8vxwkYKpU74Gcy1BVg",
    "vM0Rtnrpx4aDVWy4eFO8vAUuTf",
    "wgLgrNpNO6WYMjMSbiGiI4ABLt",
  ],
  "user": Object {
    "description": "Extended health coverage for adults with low incomes",
    "detailLinks": Array [],
    "details": "
What is it?
The Alberta Adult Health Benefit (AAHB) program supports Albertans with limited incomes by covering the cost of prescription drugs, dental services, optical services, emergency ambulance, and diabetic supplies. The AAHB pays for supplies that are not available through standard Alberta health coverage.

Who is eligible?
Residents of Alberta are eligible for the AAHB if they meet the income criteria as well as one of the following conditions:
- Leaving Alberta Works Income Support or Assured Income for the Severely Handicapped (AISH) and has income from employment, self-employment, or the Canada Pension Plan Disability program
- High ongoing prescription drug needs
- Pregnant

An individual or family is NOT eligible for AAHB if they are receiving extended health benefits from other government programs including: 
- Income Support/AISH
- Child and Youth Supports (Child Health Benefit)
- Government-funded Indigenous health programs
- Sponsored immigrants or refugees
- Seniors (65+) health programs

What do I need to apply?
- SIN numbers for all family members
- Alberta Personal Health Number
- Completed taxes so Alberta Health Services can check your income

For more information and to find the application follow the link below.",
    "externalLink": "www.alberta.ca/alberta-adult-health-benefit.aspx",
    "guid": "hp5mupqcVY8ZMiKg7Q91Uoi4Wf",
    "tags": Array [
      "Health",
      "Alberta",
    ],
    "title": "Adult Health Benefit",
  },
}
`;
